### PR TITLE
doc: devicetree: Document the define generation rules

### PR DESCRIPTION
**Pull request by galak**
Starts of my attempt at documenting how / what we generate w/regards to DT defines.

### CODE REVIEW

**Comment by galak** [comment]
Here are my notes for node property defines:
```
PROPERTY:
* reg / reg-names
* irqs
* general props:
  - bool
  #define DT_FIXED_PARTITIONS_20000_READ_ONLY         0 
  - int
  #define DT_ST_STM32_PWM_40000000_PWM_ST_PRESCALER   0
  - string
  #define DT_ST_STM32_TIMERS_40000000_LABEL           "TIMERS_2"

  - 'array' [FROM efr32mg_sltb004a]
#define DT_SILABS_GECKO_I2C_4000C400_LOCATION_SCL_0 17
#define DT_SILABS_GECKO_I2C_4000C400_LOCATION_SCL_1 2
#define DT_SILABS_GECKO_I2C_4000C400_LOCATION_SCL_2 5
#define DT_SILABS_GECKO_I2C_4000C400_LOCATION_SCL   {17, 2, 5}  	
  - 'string-array'
+#define DT_ST_STM32_I2C_V2_40005C00_INTERRUPT_NAMES_0 "event"
+#define DT_ST_STM32_I2C_V2_40005C00_INTERRUPT_NAMES_1 "error"
  - 'uint8-array'
  [ FROM REELBOARD ]
#define DT_NORDIC_NRF_SPI_40004000_SOLOMON_SSD16XXFB_0_SDV {0x41, 0xa8, 0x32}

  - 'phandle-array'
   	
  - enum:
  +#define DT_ST_STM32_OTGFS_50000000_MAXIMUM_SPEED_ENUM 1

* clocks


* spi-dev (CS-GPIO)

+#define DT_ST_STM32_SPI_FIFO_40003C00_ZEPHYR_BT_HCI_SPI_0_CS_GPIOS_CONTROLLER "GPIOD"
+#define DT_ST_STM32_SPI_FIFO_40003C00_ZEPHYR_BT_HCI_SPI_0_CS_GPIOS_PIN 13
+#define DT_ST_STM32_SPI_FIFO_40003C00_ZEPHYR_BT_HCI_SPI_0_CS_GPIOS_FLAGS 16
+#define DT_ST_STM32_SPI_FIFO_40003C00_ZEPHYR_BT_HCI_SPI_0_CS_GPIOS {"GPIOD", 13, 16}


* bus-name
#define DT_ST_STM32_SPI_FIFO_40003C00_ZEPHYR_BT_HCI_SPI_0_BUS_NAME "SPI_3"
```

File: doc/guides/dts/index.rst [hunk 1]
```diff
@@ -398,6 +398,91 @@ usually take a ``struct device*`` as their first argument. This allows the
 driver API to use information from devicetree to interact with the device
 hardware.
 
+Generated macro rules
+=====================
+.. code-block:: none
+
+      aliases {
```
Comment (ulfalizer) [inline doc/guides/dts/index.rst]: Inconsistently indented with the table below.

File: doc/guides/dts/index.rst [hunk 2]
```diff
@@ -398,6 +398,91 @@ usually take a ``struct device*`` as their first argument. This allows the
...
+     -
+     - ``#define DT_CCM_...``
+   * - memory device
+     - ``#define DT_DTCM_...``
+     -
+     - ``#define DT_DTCM_...``
+   * - memory device
+     - ``#define DT_IPC_SHM_...``
+     -
+     - ``#define DT_IPC_SHM_...``
```
Comment (ulfalizer) [inline doc/guides/dts/index.rst]: Some descriptions and examples in plain English would be nice. Having a table as an overview might be a good idea still.

Should explain how the devicetree source is related to the table at least, to tie them together.

File: doc/guides/dts/index.rst [hunk 3]
```diff
@@ -398,6 +398,91 @@ usually take a ``struct device*`` as their first argument. This allows the
...
+                     compatible = "nxp,fxos8700";
+                     reg = <0x1d>;
+                };
+           };
+      };
+
+.. list-table::
+   :header-rows: 1
+
+   * - type
```
Comment (ulfalizer) [inline doc/guides/dts/index.rst]: Could capitalize the headings, etc. Nice to aim for more "writing class" than "engineering class" in the documentation.

**Comment by ulfalizer** [comment]
Working on documenting the basic ``DT_<node>_<property>`` format completely now. Think I can finally remember how it works after the work on replacing those deprecated aliases.

Wasn't triggered by this PR. Just happened to be working on a similar thing. Might not overlap.

File: doc/guides/dts/index.rst [hunk 4]
```diff
@@ -398,6 +398,91 @@ usually take a ``struct device*`` as their first argument. This allows the
...
+   * - node
+     - ``DT_ALIAS_<ALIAS>_...``
+     -
+     - ``DT_ALIAS_I2C_0_...``
+   * - node
+     - ``DT_<COMPAT>_<ALIAS>_...``
+     -
+     - ``DT_NXP_KINETIS_I2C_I2C_0_...``
+   * - existence
+     - ``#define DT_COMPAT_<COMPAT> 1``
```
Comment (carlescufi) [inline doc/guides/dts/index.rst]: well, can we get rid of the `DT_COMPAT_<COMPAT>_` and `DT_<COMPAT>_BUS_` if we have `DT_INST_`?

**Review by carlescufi — COMMENTED** [review]
Looks good, just a question.

Also, I think we should have a PR that defines the overall DT goals within Zephyr **before** we review this one.
